CompuCell3D/core/CompuCell3D/steppables/TubeFieldInitializer/TubeFieldInitializer.cpp

```diff
@@ -0,0 +1,303 @@
+
+    Dim3D dim = cellField->getDim();
+
+    cerr << "dim:" << dim.x << ", " << dim.y << ", "  << dim.z << endl; 
+	Dim3D itDim = getTubeDimensions(dim, size);
+    cerr << "itDim:" << itDim.x << ", " << itDim.y << ", "  << itDim.z << endl;
+	CC3D_Log(LOG_DEBUG) << "itDim="<<itDim;
+
+    Point3D pt;
+    Point3D cellPt;
+    CellG *cell;
+
+    double tubeLength = distance(_initData.fromPoint.x, _initData.fromPoint.y, _initData.fromPoint.z,
+            _initData.toPoint.x, _initData.toPoint.y, _initData.toPoint.z);
+    double hypotenuse = sqrt( pow(_initData.outerRadius, 2) + pow(tubeLength, 2) );
+
+    for (int z = 0; z < itDim.z; z++)
+        for (int y = 0; y < itDim.y; y++)
+            for (int x = 0; x < itDim.x; x++) {
+                pt.x = x * size;
+                pt.y = y * size;
+                pt.z = z * size;
+
+                //Step 1: Is the point close/far enough to the center axis of the tube?
+                double dist = distanceToLine(_initData.fromPoint, _initData.toPoint, pt);
+                if (dist > _initData.outerRadius || dist < _initData.innerRadius) {
+                    continue;
+                }
+
+                //Step 2: Is the point too far from the face of the tube/cylinder? (Pythagorean Thm.)
+                //This trims the tube to its desired length.
+                //1. Choose the face that's farther away from the point pt.
+                double fromDist = distance(_initData.fromPoint.x, _initData.fromPoint.y, _initData.fromPoint.z,
+                        pt.x, pt.y, pt.z);
+                double toDist = distance(_initData.toPoint.x, _initData.toPoint.y, _initData.toPoint.z,
+                        pt.x, pt.y, pt.z);
+                double distanceFromFace = max(fromDist, toDist);
+                //2. Check the distance from `pt` to the farther face against hypotenuse.
+                //Ex: If they are equal, the point would be at the bottom edge of the tube.
+                //Ex: If the dist is too great, the point is beyond the tube's length. 
+                if (distanceFromFace > hypotenuse) {
+                    continue;
+                }
+
+
+                if (BoundaryStrategy::getInstance()->isValid(pt)) {
+                    cell = potts->createCellG(pt);
+                    cell->type = initCellType(_initData);
+                    potts->runSteppers(); //used to ensure that VolumeTracker Plugin step fcn gets called every time we do something to the fields
+                    //It is necessary to do it this way because steppers are called only when we are performing pixel copies
+                    // but if we initialize steppers are not called thus is you overwrite a cell here it will not get removed from
+                    //inventory unless you call steppers(VolumeTrackerPlugin) explicitly
+
+
+                } else {
+                    continue;
+                }
+
+                for (cellPt.z = pt.z; cellPt.z < pt.z + cellWidth &&
```

Comment:
same comment as on another PR - let's create a function that lays out cells and put the 3 nested loops there